### ISSUE TITLE
Validating upload/download tokens in Gatekeeper app

### ISSUE THREAD

**Issue by cameel** [golemfactory/concent#7]
Extend the Gatekeeper app (#6) to add more detailed logic for processing tokens passed to it in the `Authorization` header.

### Token structure
Token is a type of a Golem message (let's call it `ConcentFileTransferToken`). It has the following fields

| field                                  | type                                    | description |
|----------------------------------------|-----------------------------------------|-------------|
| `token_expiration_deadline`            | int (UNIX timestamp)                    | The timestamp that determines the maximum time to sumit the token. After that point in time the gatekeeper unconditionally rejects the token as invalid.
| `storage_cluster_address`              | string (absolute URL)                   | The address of the cluster that stores files.
| `authorized_client_public_key`         | bytes (ECDSA public key)                | Public key of the Golem client authorized to transfer the files.
| `files`                                | list of string (list of relative paths) | The list of files the client is authorized to transfer. Relative to a directory defined in server configuration.
| `operation`                            | string (`'upload'` or `'download'`)         | The operation the client is authorized to perform.

The message is not yet defined in `golem-messages`. Please use a custom fork of that library for development and add the message type there. We're going to resolve this issue before the token validation feature is finalized.

### Token validation
A token should be accepted if: (1) the request is well-formed and (2) the token is well-formed and (3) the token is valid for the requested transfer.

A gatekeeper request is well-formed when:
- If it's an upload request:
    - The method is POST.
    - It's a `multipart/form-data` request.
    - Exactly one file is being uploaded.
    - The file path (the part of the URL that follows `/gatekeeper/upload-auth/`) is non-empty and contains a relative path.
- If it's a download request:
    - The method is GET.
    - The request has no body and `Content-Type` is not specified.
    - The file path (the part of the URL that follows `/gatekeeper/upload-auth/`) is non-empty and contains a relative path.
- The request does not have headers called `Concent-Upload-Path` or `X-Session-ID`
    - This is temporary, just in case something gets through from our nginx due to a bug.

A well-formed token must satisfy the following conditions:
- All fields are non-`null`.
- Values of all fields match their types.
- `storage_cluster_address` is a valid URL (use Django's [URLValidator](https://docs.djangoproject.com/en/1.11/ref/validators/#urlvalidator)).

A token is valid for a specific transfer when:
- current time <= `token_expiration_deadline`
- `storage_cluster_address` matches the value of a setting called `STORAGE_CLUSTER_ADDRESS`.
- `authorized_client_public_key` matches the key sent by the client in the `Concent-Client-Public-Key` HTTP header.
- The value of the `operation` field matches the operation the client is trying to perform.
- If `operation` is `'upload'`:
    - The file path sent by the client is present on the `files` list.
- If `operation` is `'download'`:
    - The file path sent by the client is present on the `files` list.

### `STORAGE_CLUSTER_ADDRESS` setting
Add a setting called `STORAGE_CLUSTER_ADDRESS`. Add a [Django check](https://docs.djangoproject.com/en/2.0/topics/checks/) that ensures that it satisfies the following conditions:
- It's defined if `'gatekeeper'` is in `CONCENT_FEATURES`.
- Uses `http` or `https` protocol.
- It is a valid URL.

The setting should be undefined by default. For development (in `development.py`) set it to `http://localhost:8000/`

### Other
- Ignore the file name specified in `Content-Disposition` of a multipart request. It can't contain directory names and this is a necessary feature for us.
- For now we trust that the public key provided by the client belongs to it. Validating this will be the subject of another task in the future (it's not specified yet how we're going to do this).

#### Download
The request that goes to `nginx-storage`:
``` bash
curl http://storage.concent.golem.network/download/blender/benchmark/test_task/scene-Helicopter-27-cycles.blend \
    --header "Authorization: Golem dG9rZW4K" \
    --header "Concent-Client-Public-Key: MTIzNDU2Cg=="
```

The request equivalent to the one `nginx-storage` sends to `gatekeeper`:
``` bash
curl http://gatekeeper-container/gatekeeper/download-auth/blender/benchmark/test_task/scene-Helicopter-27-cycles.blend \
    --header "Authorization: Golem dG9rZW4K" \
    --header "Concent-Client-Public-Key: MTIzNDU2Cg=="
```

The values in the token:

| field                                  | value
|----------------------------------------|---------------------------|
| `token_expiration_deadline`            | `1512122400` (2017-12-01 11:00)
| `cluster_address`                      | `http://storage.concent.golem.network/`
| `authorized_client_public_key`         | `123456`
| `files`                                | `['blender/benchmark/test_task/scene-Helicopter-27-cycles.blend', 'blender/benchmark/test_task/texture.tiff']`
| `operation`                            | `'download'`

#### Upload
The request that goes to `nginx-storage`:
``` bash
curl -X POST http://storage.concent.golem.network/upload/blender/benchmark/test_task/scene-Helicopter-27-cycles.blend \
    --header "Authorization: Golem dG9rZW4K" \
    --header "Concent-Client-Public-Key: MTIzNDU2Cg==" \
    --form file=scene-Helicopter-27-cycles.blend
```

The request equivalent to the one `nginx-storage` sends to `gatekeeper`:
``` bash
curl -X POST http://gatekeeper-container/gatekeeper/upload-auth/blender/benchmark/test_task/scene-Helicopter-27-cycles.blend \
    --header "Authorization: Golem dG9rZW4K" \
    --header "Concent-Client-Public-Key: MTIzNDU2Cg=="
```

Note that gatekeeper does not get the request body.

The values in the token:

| field                                  | value
|----------------------------------------|---------------------------|
| `token_expiration_deadline`            | `1512122400` (2017-12-01 11:00)
| `cluster_address`                      | `http://storage.concent.golem.network/`
| `authorized_client_public_key`         | `123456`
| `files`                                | `['blender/benchmark/test_task/scene-Helicopter-27-cycles.blend', 'blender/benchmark/test_task/texture.tiff']`
| `operation`                            | `'upload'`

**Comment by cameel** [comment]
Task updated:
- Now path of the file to be uploaded is part of the request path, both for upload and download
- `Concent-Upload-Path` or `X-Session-ID` HTTP headers should not be used.
- Added form data to the upload example.

**Comment by Jakub89** [comment]
> - `authorized_client_public_key` matches the key sent by the client in the `Concent-Client-Public-Key` HTTP header.

Should we check something more about those keys? Because right now they can be empty strings, and validation will pass.

**Comment by cameel** [comment]
Well, it's a good idea to check that the key is present and that it's not empty. But it's a job for the `api_view` decorator. In the view you can assume that if it passed through the decorator and satisfies all conditions, the key is fine.

But this will never be foolproof and is not meant to be. The assumption is that the Concent API server will never give you a token with invalid `authorized_client_public_key` inside. And writing the code that generates the token is your next task (but it's still waiting for the spec from me). This will be the right place for more checks.

And also, as stated in the description, in this version we don't really ensure that the key is valid and that we can trust it. It's a part that still needs some design work between us and Golem and I'm leaving it out of this task for the time being.

**Comment by Jakub89** [comment]
I've switched 'multipart/form-data' request checks to 'application/x-www-form-urlencoded', because of nginx dependency. So now i upload curl we need for example `--data-binary "@scene-Helicopter-27-cycles.blend"` instead of  `--form file=scene-Helicopter-27-cycles.blend`.

**Comment by cameel** [comment]
A small clarification:
- `STORAGE_CLUSTER_ADDRESS` setting is required only if `'gatekeeper'` is in `CONCENT_FEATURES`.

**Comment by cameel** [comment]
Merged into master: 5adb018be5fdc118368a8bb80c355d7bdaed88a3
